File: Python3/aoc2020/solutions/day_20.py
from __future__ import annotations

from collections import defaultdict
from functools import reduce
from operator import mul
from typing import Dict, Iterator, List, Sequence, Set, Tuple, TypeVar

from .solution import Solution

TGridItem = TypeVar("TGridItem")
Grid = List[List[TGridItem]]

Borders = Tuple[int, int, int, int]


def border_value(border: Sequence[str]) -> int:
    return reduce(lambda acc, v: 2 * acc + (v == "#"), reversed(border), 0)


class Image:
    __slots__ = "image"

    image: Grid[str]

    def __init__(self, image: Grid[str]):
        assert len(image) == len(image[0])

        self.image = image

    def __getitem__(self, key):
        return self.image[key]

    def __str__(self) -> str:
        return "\n".join("".join(row) for row in self.image)

    @property
    def size(self) -> int:
        return len(self.image)

    def rotate_clockwise(self) -> Image:
        return Image([list(row) for row in zip(*self.image)])

    def flip_vertically(self) -> Image:
        return Image([row[::-1] for row in self.image])

    def flip_horizontally(self) -> Image:
        return Image([row[:] for row in reversed(self.image)])

    def clone(self) -> Image:
        return Image([row[:] for row in self.image])

    def borders(self) -> Borders:
        return (
            border_value(self.image[0]),
            border_value(list(zip(*self.image))[-1]),
            border_value(self.image[-1]),
            border_value(next(zip(*self.image))),
        )

    def possible_images(self) -> Iterator[Image]:
        for base_configuration in [
            self.clone(),
            self.flip_vertically(),
            self.flip_horizontally(),
            self.flip_horizontally().flip_vertically()
        ]:
            yield base_configuration

            for _ in range(3):
                base_configuration = base_configuration.rotate_clockwise()
                yield base_configuration

    @classmethod
    def square(cls, size: int) -> Image:
        return cls([["."] * size for _ in range(size)])


class Tile:
    __slots__ = "id", "image"

    def __init__(self, _id: int, image: Image):
        self.id = _id
        self.image = image

    @property
    def size(self):
        return self.image.size

    def unique_border_sides(self):
        return set(b for image in self.possible_images() for b in image.borders())

    def possible_images(self) -> Iterator[Image]:
        return self.image.possible_images()

    def __repr__(self):
        return f"Tile {self.id}"

    def __str__(self):
        return f"Tile {self.id}"

    def __hash__(self):
        return hash(self.id)

    def __eq__(self, other):
        return isinstance(other, Tile) and self.id == other.id

    @ classmethod
    def from_str(cls, tile_text: str) -> Tile:
        lines = tile_text.splitlines()
        id_value = int(lines[0][-5: -1], base=10)
        image = Image([list(lines[idx]) for idx in range(1, len(lines))])

        return cls(id_value, image)


ArrangementGraph = Dict[Tile, Dict[Tile, List[int]]]
Border2Tiles = Dict[int, Set[Tile]]


class Arrangement:
    __slots__ = "size", "grid"

    size: int
    grid: Grid[Tile]

    def __init__(self, size: int):
        self.size = size
        self.grid = []

    def collect_border2tiles(self, tiles: List[Tile]) -> Border2Tiles:
        border2tiles = defaultdict(set)

        for tile in tiles:
            for side in tile.unique_border_sides():
                border2tiles[side].add(tile)

        return border2tiles

    def recover_tile_graph(self, border2tiles: Border2Tiles) -> ArrangementGraph:
        graph: ArrangementGraph = defaultdict(lambda: defaultdict(list))
        for side, tiles in border2tiles.items():
            if len(tiles) == 1:
                continue

            assert len(tiles) == 2

            first, second = list(tiles)
            graph[first][second].append(side)
            graph[second][first].append(side)

        for edges in graph.values():
            assert 2 <= len(edges) <= 4

        return graph

    def _find_corner_tiles(self, border2tiles: Border2Tiles) -> List[Tile]:
        counter: Dict[Tile, int] = defaultdict(int)

        for side, tiles in border2tiles.items():
            if len(tiles) == 1:
                continue

            for tile in tiles:
                counter[tile] += 1

        for tile in counter:
            counter[tile] //= 2

        border_tiles = [tile for tile,
                        neighbours in counter.items() if neighbours == 2]

        assert len(border_tiles) == 4

        return border_tiles

    def recover_arrangement_grid(self, starting_tile: Tile, graph: ArrangementGraph) -> Grid[Tile]:
        visited = set([starting_tile])
        arrangement = []
        current_row = [starting_tile]
        for _ in range(self.size - 1):
            for neighbour in graph[current_row[-1]]:
                if neighbour not in visited and len(graph[neighbour]) < 4:
                    visited.add(neighbour)
                    current_row.append(neighbour)
                    break

        arrangement.append(current_row[:])

        for _ in range(self.size - 1):
            current_row.clear()

            for last_elem in arrangement[-1]:
                for n in graph[last_elem]:
                    if n not in visited:
                        visited.add(n)
                        current_row.append(n)
                        break

            arrangement.append(current_row[:])

        for r in range(self.size):
            for c in range(self.size):
                tile = arrangement[r][c]

                tile_orientation_found = False
                for possible_image in tile.possible_images():
                    for border, (dr, dc) in enumerate(((-1, 0), (0, 1), (1, 0), (0, -1))):
                        nr, nc = r + dr, c + dc
                        if nr < 0 or nr >= self.size or nc < 0 or nc >= self.size:
                            continue

                        possible_border = possible_image.borders()[border]
                        allowed_borders = graph[tile][arrangement[nr][nc]]

                        if possible_border not in allowed_borders:
                            break
                    else:
                        arrangement[r][c].image = possible_image
                        tile_orientation_found = True
                        break

                assert tile_orientation_found

        self.validate_arrangement_grid(arrangement)

        return arrangement

    def validate_arrangement_grid(self, arrangement: Grid[Tile]):
        for r in range(self.size):
            for c in range(self.size):
                tile = arrangement[r][c]
                tile_border = tile.image.borders()
                for border, (dr, dc) in enumerate(((-1, 0), (0, 1), (1, 0), (0, -1))):
                    nr, nc = r + dr, c + dc
                    if nr < 0 or nr >= self.size or nc < 0 or nc >= self.size:
                        continue
                    n_border = arrangement[nr][nc].image.borders()

                    assert tile_border[border] == n_border[(border + 2) % 4]

    def recover_from_tiles(self, tiles: List[Tile]) -> None:
        border2tiles = self.collect_border2tiles(tiles)
        graph = self.recover_tile_graph(border2tiles)

        assert len(graph) == len(tiles)

        border_tiles = self._find_corner_tiles(border2tiles)
        grid = self.recover_arrangement_grid(border_tiles[0], graph)

        self.grid = grid

    def find_corner_tiles(self) -> List[Tile]:
        return [
            self.grid[0][0],
            self.grid[0][-1],
            self.grid[-1][-1],
            self.grid[-1][0],
        ]

    def recover_original_image(self) -> Image:
        tile_size = self.grid[0][0].size
        big_tile_size = self.size * (tile_size - 2)
        big_tile_image = Image.square(big_tile_size)

        for row, grid_row in enumerate(self.grid):
            for col, tile in enumerate(grid_row):
                for tr, tile_row in enumerate(tile.image[1:-1]):
                    for tc, tile_col in enumerate(tile_row[1:-1]):
                        nr = row * (tile_size - 2) + tr
                        nc = col * (tile_size - 2) + tc
                        big_tile_image[nr][nc] = tile_col

        return big_tile_image

    def __str__(self):
        tile_size = self.grid[0][0].size
        screen = []

        for row, grid_row in enumerate(self.grid):
            line = [[] for _ in range(tile_size)]

            for col, tile in enumerate(grid_row):

                for i, r in enumerate(tile.image):
                    line[i].append("".join(r))

            for l in line:
                screen.append(" ".join(l))

            screen.append("")

        return "\n".join(screen)

    @ classmethod
    def recover_from_str(cls, tiles_text: str) -> Arrangement:
        tiles = [Tile.from_str(tile_text)
                 for tile_text in tiles_text.split("\n\n")]

        arrangement = cls(int(pow(len(tiles), 0.5)))
        arrangement.recover_from_tiles(tiles)

        return arrangement


def has_dragon_at(image: Image, r: int, c: int) -> bool:
    for dr, dc in dragon_points():
        nr, nc = r + dr, c + dc
        if nr >= image.size or nc >= image.size:
            return False

        if image.image[nr][nc] != "#":
            return False
    return True


def mark_dragon_at(image: Image, r: int, c: int) -> None:
    for dr, dc in dragon_points():
        nr, nc = r + dr, c + dc

        if nr >= image.size or nc >= image.size:
            continue

        image.image[nr][nc] = "O"


def dragon_points():
    dragon = ["                  # ",
              "#    ##    ##    ###",
              " #  #  #  #  #  #   "]

    for r, row in enumerate(dragon):
        for c, v in enumerate(row):
            if v == "#":
                yield r, c


def try_mark_sea_monsters(image: Image) -> int:
    size = image.size
    count = 0

    for r in range(size):
        for c in range(size):
            if has_dragon_at(image, r, c):
                count += 1
                mark_dragon_at(image, r, c)

    return count


def mark_image_with_sea_monsters(image: Image) -> Image:
    for possible_image in image.possible_images():
        if try_mark_sea_monsters(possible_image) > 0:
            return possible_image

    return image


def count_hashes_in(image: Image) -> int:
    hashes = 0
    for r, row in enumerate(image.image):
        for c, v in enumerate(row):
            if v == "#":
                hashes += 1
    return hashes


def count_product_of_corners_in_arrangement(arrangement: Arrangement):
    border_tiles = arrangement.find_corner_tiles()

    return reduce(mul, (tile.id for tile in border_tiles), 1)


def count_water_roughness(arrangement: Arrangement):
    recovered_image = arrangement.recover_original_image()
    marked_image = mark_image_with_sea_monsters(recovered_image)

    return count_hashes_in(marked_image)


class Day20(Solution):

    def first_task(self, tiles_text: str) -> str:
        arrangement = Arrangement.recover_from_str(tiles_text)

        return str(count_product_of_corners_in_arrangement(arrangement))

    def second_task(self, tiles_text: str) -> str:
        arrangement = Arrangement.recover_from_str(tiles_text)

        return str(count_water_roughness(arrangement))

File: Rust/src/solutions/day_20.rs
use super::solution::{Error, Solution};
use std::cmp::{Eq, PartialEq};
use std::collections::{HashMap, HashSet};
use std::hash::{Hash, Hasher};

type Border = usize;
type Borders = [Border; 4];
type Grid<T> = Vec<Vec<T>>;
type ArrangementGraph<'a> = HashMap<&'a Tile, HashMap<&'a Tile, Vec<Border>>>;
type Border2Tiles<'a> = HashMap<Border, HashSet<&'a Tile>>;

fn border_value(border: &[char]) -> usize {
    border
        .iter()
        .rev()
        .fold(0, |acc, &el| 2 * acc + if el == '#' { 1 } else { 0 })
}

struct Image {
    image: Grid<char>,
}

fn create_square_grid(size: usize) -> Grid<char> {
    vec![vec!['.'; size]; size]
}

impl Image {
    fn from_str(s: &str) -> Image {
        Image {
            image: s.lines().map(|l| l.chars().collect::<Vec<_>>()).collect(),
        }
    }
    fn square(size: usize) -> Image {
        Image {
            image: create_square_grid(size),
        }
    }

    fn size(&self) -> usize {
        self.image.len()
    }

    fn borders(&self) -> Borders {
        [
            border_value(&self.image[0]),
            border_value(
                &(0..self.size())
                    .map(|r| self.image[r][self.size() - 1])
                    .collect::<Vec<_>>(),
            ),
            border_value(&self.image[self.size() - 1]),
            border_value(
                &(0..self.size())
                    .map(|r| self.image[r][0])
                    .collect::<Vec<_>>(),
            ),
        ]
    }

    fn clone(&self) -> Image {
        let mut image = create_square_grid(self.size());

        for (r, row) in image.iter_mut().enumerate().take(self.size()) {
            for (c, chr) in row.iter_mut().enumerate().take(self.size()) {
                *chr = self.image[r][c];
            }
        }

        Image { image }
    }

    fn rotate_clockwise(&self) -> Image {
        let mut image = create_square_grid(self.size());

        for (r, row) in image.iter_mut().enumerate().take(self.size()) {
            for (c, chr) in row.iter_mut().enumerate().take(self.size()) {
                *chr = self.image[self.size() - c - 1][r];
            }
        }

        Image { image }
    }

    fn flip_horizontally(&self) -> Image {
        let mut image = create_square_grid(self.size());

        for (r, row) in image.iter_mut().enumerate().take(self.size()) {
            for (c, chr) in row.iter_mut().enumerate().take(self.size()) {
                *chr = self.image[self.size() - r - 1][c];
            }
        }

        Image { image }
    }

    fn flip_vertically(&self) -> Image {
        let mut image = create_square_grid(self.size());

        for (r, row) in image.iter_mut().enumerate().take(self.size()) {
            for (c, chr) in row.iter_mut().enumerate().take(self.size()) {
                *chr = self.image[r][self.size() - c - 1];
            }
        }

        Image { image }
    }

    fn possible_images(&self) -> Vec<Image> {
        let mut images = Vec::new();

        for base_image in &[
            self.clone(),
            self.flip_horizontally(),
            self.flip_vertically(),
            self.flip_horizontally().flip_vertically(),
        ] {
            let mut base_image = base_image.clone();
            for _ in 0..4 {
                images.push(base_image.rotate_clockwise());
                base_image = base_image.rotate_clockwise();
            }
        }

        images
    }
}

struct Tile {
    id: usize,
    image: Image,
}

impl PartialEq for Tile {
    fn eq(&self, other: &Tile) -> bool {
        self.id == other.id
    }
}

impl Eq for Tile {}

impl Hash for Tile {
    fn hash<H>(&self, state: &mut H)
    where
        H: Hasher,
    {
        self.id.hash(state)
    }
}

impl Tile {
    fn size(&self) -> usize {
        self.image.size()
    }

    fn unique_border_sides(&self) -> Vec<usize> {
        self.image
            .possible_images()
            .iter()
            .flat_map(|img| img.borders().iter().copied().collect::<Vec<usize>>())
            .collect::<HashSet<usize>>()
            .iter()
            .copied()
            .collect::<Vec<usize>>()
    }

    fn possible_images(&self) -> Vec<Image> {
        self.image.possible_images()
    }

    fn from_str(tile_text: &str) -> Tile {
        let mut lines = tile_text.lines();
        let id_value = lines
            .next()
            .unwrap()
            .trim_start_matches("Tile ")
            .trim_end_matches(':')
            .parse::<usize>()
            .unwrap();

        let image_text = lines.collect::<Vec<&str>>().join("\n");
        let image = Image::from_str(&image_text);

        Tile {
            id: id_value,
            image,
        }
    }
}

struct Arrangement {
    size: usize,
    grid: Grid<Tile>,
}

fn collect_border2tiles(tiles: &[Tile]) -> Border2Tiles {
    let mut border2tiles: Border2Tiles = HashMap::new();

    for tile in tiles {
        for border in tile.unique_border_sides() {
            let entry = border2tiles.entry(border).or_insert_with(HashSet::new);

            entry.insert(&tile);
        }
    }

    border2tiles
}

fn recover_tile_graph<'a>(border2tiles: &'a Border2Tiles) -> ArrangementGraph<'a> {
    let mut graph: ArrangementGraph = HashMap::new();

    for (side, tiles) in border2tiles {
        if tiles.len() == 1 {
            continue;
        }

        let tiles_list = tiles.iter().collect::<Vec<_>>();
        let first = tiles_list[0];
        let second = tiles_list[1];

        let first_entry = graph
            .entry(*first)
            .or_insert_with(HashMap::new)
            .entry(*second)
            .or_insert(vec![]);
        first_entry.push(*side);

        let second_entry = graph
            .entry(*second)
            .or_insert_with(HashMap::new)
            .entry(*first)
            .or_insert(vec![]);
        second_entry.push(*side);
    }

    graph
}

fn find_corner_tiles<'a>(border2tiles: &'a Border2Tiles) -> Vec<&'a Tile> {
    let mut counter: HashMap<&Tile, usize> = HashMap::new();

    for tiles in border2tiles.values() {
        if tiles.len() == 1 {
            continue;
        }

        for tile in tiles {
            *counter.entry(tile).or_insert(0) += 1;
        }
    }

    counter
        .iter()
        .filter(|(_, &size)| size == 4)
        .map(|(&tile, _)| tile)
        .collect::<Vec<_>>()
}

fn recover_arrangement_grid<'a>(
    size: usize,
    starting_tile: &'a Tile,
    graph: &'a ArrangementGraph,
) -> Grid<Tile> {
    let mut visited: HashSet<&Tile> = HashSet::new();
    visited.insert(starting_tile);

    let mut arrangement: Grid<&Tile> = vec![];
    let mut current_row: Vec<&Tile> = vec![starting_tile];

    for _ in 0..size - 1 {
        for neighbour in graph[current_row[current_row.len() - 1]].keys() {
            if !visited.contains(neighbour) && graph[neighbour].len() < 4 {
                visited.insert(neighbour);
                current_row.push(neighbour);
                break;
            }
        }
    }

    arrangement.push(current_row);

    for _ in 0..size - 1 {
        let mut current_row = vec![];

        for &last_elem in arrangement[arrangement.len() - 1].iter() {
            for &n in graph[last_elem].keys() {
                if !visited.contains(n) {
                    visited.insert(n);
                    current_row.push(n);
                    break;
                }
            }
        }

        arrangement.push(current_row);
    }

    let mut fixed_arrangement: Grid<Tile> = vec![];

    for r in 0..size {
        let mut current_row = vec![];

        for c in 0..size {
            let tile = arrangement[r][c];

            for possible_image in tile.possible_images() {
                let mut orientation_found = true;

                for (border, (dr, dc)) in [(-1_i32, 0), (0, 1), (1, 0), (0, -1_i32)]
                    .iter()
                    .enumerate()
                {
                    let nr: i32 = r as i32 + dr;
                    let nc: i32 = c as i32 + dc;

                    if nr < 0 || nr as usize >= size || nc < 0 || nc as usize >= size {
                        continue;
                    }

                    let possible_border = possible_image.borders()[border];
                    let allowed_borders = graph
                        .get(tile)
                        .unwrap()
                        .get(arrangement[nr as usize][nc as usize])
                        .unwrap();

                    if !allowed_borders.contains(&possible_border) {
                        orientation_found = false;
                        break;
                    }
                }

                if orientation_found {
                    current_row.push(Tile {
                        id: tile.id,
                        image: possible_image,
                    });

                    break;
                }
            }
        }

        fixed_arrangement.push(current_row);
    }

    validate_arrangement_grid(size, &fixed_arrangement);

    fixed_arrangement
}

fn validate_arrangement_grid(size: usize, arrangement: &Grid<Tile>) {
    for r in 0..size {
        for c in 0..size {
            let tile = &arrangement[r][c];
            let tile_border = tile.image.borders();

            for (border, (dr, dc)) in [(-1_i32, 0), (0, 1), (1, 0), (0, -1_i32)]
                .iter()
                .enumerate()
            {
                let nr: i32 = r as i32 + dr;
                let nc: i32 = c as i32 + dc;

                if nr < 0 || nr as usize >= size || nc < 0 || nc as usize >= size {
                    continue;
                }

                let n_border = &arrangement[nr as usize][nc as usize].image.borders();

                assert!(tile_border[border] == n_border[(border + 2) % 4]);
            }
        }
    }
}

fn recover_from_tiles(size: usize, tiles: Vec<Tile>) -> Grid<Tile> {
    let border2tiles = collect_border2tiles(&tiles);
    let graph = recover_tile_graph(&border2tiles);
    let border_tiles = find_corner_tiles(&border2tiles);

    recover_arrangement_grid(size, border_tiles[0], &graph)
}

impl Arrangement {
    fn find_corner_tiles(&self) -> [usize; 4] {
        [
            self.grid[0][0].id,
            self.grid[0][self.grid.len() - 1].id,
            self.grid[self.grid.len() - 1][self.grid.len() - 1].id,
            self.grid[self.grid.len() - 1][0].id,
        ]
    }

    fn recover_original_image(&self) -> Image {
        let tile_size = self.grid[0][0].size();
        let big_tile_size = self.size * (tile_size - 2);
        let mut big_tile_image = Image::square(big_tile_size);

        for (row, grid_row) in self.grid.iter().enumerate() {
            for (col, tile) in grid_row.iter().enumerate() {
                for tr in 1..(tile_size - 1) {
                    for tc in 1..(tile_size - 1) {
                        let nr = row * (tile_size - 2) + tr - 1;
                        let nc = col * (tile_size - 2) + tc - 1;
                        big_tile_image.image[nr][nc] = tile.image.image[tr][tc];
                    }
                }
            }
        }

        big_tile_image
    }

    fn from_str(tiles_text: &str) -> Arrangement {
        let tiles = tiles_text
            .split("\n\n")
            .map(|t| Tile::from_str(t))
            .collect::<Vec<_>>();

        let size = (tiles.len() as f64).sqrt() as usize;
        let grid = recover_from_tiles(size, tiles);

        Arrangement { size, grid }
    }
}

fn count_product_of_corners_in_arrangement(arr: &Arrangement) -> usize {
    let border_tiles = arr.find_corner_tiles();

    border_tiles.iter().product()
}

fn dragon_points() -> Vec<(usize, usize)> {
    [
        "                  # ",
        "#    ##    ##    ###",
        " #  #  #  #  #  #   ",
    ]
    .iter()
    .enumerate()
    .flat_map(|(r, row)| row.chars().enumerate().map(move |(c, ch)| (r, c, ch)))
    .filter(|(_, _, ch)| *ch == '#')
    .map(|(r, c, _)| (r, c))
    .collect()
}

fn has_dragon_at(image: &Image, r: usize, c: usize) -> bool {
    for (dr, dc) in dragon_points() {
        let (nr, nc) = (r + dr, c + dc);
        if nr >= image.size() || nc >= image.size() {
            return false;
        }

        if image.image[nr][nc] != '#' {
            return false;
        }
    }

    true
}

fn mark_dragon_at(image: &mut Image, r: usize, c: usize) {
    for (dr, dc) in dragon_points() {
        let (nr, nc) = (r + dr, c + dc);
        if nr >= image.size() || nc >= image.size() {
            continue;
        }

        image.image[nr][nc] = 'O';
    }
}

fn try_mark_sea_monsters(image: &mut Image) -> usize {
    let size = image.size();
    let mut count = 0;

    for r in 0..size {
        for c in 0..size {
            if has_dragon_at(image, r, c) {
                count += 1;
                mark_dragon_at(image, r, c);
            }
        }
    }

    count
}

fn mark_image_with_sea_monsters(image: &Image) -> Image {
    for possible_image in image.possible_images() {
        let mut possible_image = possible_image;
        if try_mark_sea_monsters(&mut possible_image) > 0 {
            return possible_image;
        }
    }

    image.clone()
}

fn count_hashes_in(image: &Image) -> usize {
    image
        .image
        .iter()
        .flatten()
        .filter(|&&el| el == '#')
        .count()
}

fn count_water_roughness(arr: &Arrangement) -> usize {
    let recovered_image = arr.recover_original_image();
    let marked_image = mark_image_with_sea_monsters(&recovered_image);

    count_hashes_in(&marked_image)
}

pub struct Day20 {}

impl Solution for Day20 {
    fn first_task(&self, tiles_text: &str) -> Result<String, Error> {
        let arrangement = Arrangement::from_str(tiles_text);

        Ok(count_product_of_corners_in_arrangement(&arrangement).to_string())
    }

    fn second_task(&self, tiles_text: &str) -> Result<String, Error> {
        let arrangement = Arrangement::from_str(tiles_text);

        Ok(count_water_roughness(&arrangement).to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_border_function() {
        let test_borders = [
            ("#", 1),
            ("#.", 1),
            (".#", 2),
            ("#.#", 5),
            ("..##.#...", 44),
        ];

        for (border, number) in test_borders.iter() {
            assert_eq!(
                border_value(&border.chars().collect::<Vec<_>>()),
                *number as usize
            );
        }
    }

    #[test]
    fn test_borders_initial_image() {
        let test_image = "..##.
##..#
#...#
####.
##.##";

        assert_eq!(Image::from_str(test_image).borders(), [12, 22, 27, 30]);
    }

    #[test]
    fn test_borders_flip_horizontally() {
        let test_image = "..##.
##..#
#...#
####.
##.##";

        assert_eq!(
            Image::from_str(test_image).flip_horizontally().borders(),
            [27, 13, 12, 15]
        );
    }

    #[test]
    fn test_borders_flip_vertically() {
        let test_image = "..##.
##..#
#...#
####.
##.##";

        assert_eq!(
            Image::from_str(test_image).flip_vertically().borders(),
            [6, 30, 27, 22]
        );
    }

    #[test]
    fn test_first_task() {
        let test_tiles_text = "Tile 2311:
..##.#..#.
##..#.....
#...##..#.
####.#...#
##.##.###.
##...#.###
.#.#.#..##
..#....#..
###...#.#.
..###..###

Tile 1951:
#.##...##.
#.####...#
.....#..##
#...######
.##.#....#
.###.#####
###.##.##.
.###....#.
..#.#..#.#
#...##.#..

Tile 1171:
####...##.
#..##.#..#
##.#..#.#.
.###.####.
..###.####
.##....##.
.#...####.
#.##.####.
####..#...
.....##...

Tile 1427:
###.##.#..
.#..#.##..
.#.##.#..#
#.#.#.##.#
....#...##
...##..##.
...#.#####
.#.####.#.
..#..###.#
..##.#..#.

Tile 1489:
##.#.#....
..##...#..
.##..##...
..#...#...
#####...#.
#..#.#.#.#
...#.#.#..
##.#...##.
..##.##.##
###.##.#..

Tile 2473:
#....####.
#..#.##...
#.##..#...
######.#.#
.#...#.#.#
.#########
.###.#..#.
########.#
##...##.#.
..###.#.#.

Tile 2971:
..#.#....#
#...###...
#.#.###...
##.##..#..
.#####..##
.#..####.#
#..#.#..#.
..####.###
..#.#.###.
...#.#.#.#

Tile 2729:
...#.#.#.#
####.#....
..#.#.....
....#..#.#
.##..##.#.
.#.####...
####.#.#..
##.####...
##..#.##..
#.##...##.

Tile 3079:
#.#.#####.
.#..######
..#.......
######....
####.#..#.
.#...#.##.
#.#####.##
..#.###...
..#.......
..#.###...";

        let test_arrangement = Arrangement::from_str(test_tiles_text);
        assert_eq!(
            count_product_of_corners_in_arrangement(&test_arrangement),
            20899048083289
        );
    }

    #[test]
    fn test_second_task() {
        let test_tiles_text = "Tile 2311:
..##.#..#.
##..#.....
#...##..#.
####.#...#
##.##.###.
##...#.###
.#.#.#..##
..#....#..
###...#.#.
..###..###

Tile 1951:
#.##...##.
#.####...#
.....#..##
#...######
.##.#....#
.###.#####
###.##.##.
.###....#.
..#.#..#.#
#...##.#..

Tile 1171:
####...##.
#..##.#..#
##.#..#.#.
.###.####.
..###.####
.##....##.
.#...####.
#.##.####.
####..#...
.....##...

Tile 1427:
###.##.#..
.#..#.##..
.#.##.#..#
#.#.#.##.#
....#...##
...##..##.
...#.#####
.#.####.#.
..#..###.#
..##.#..#.

Tile 1489:
##.#.#....
..##...#..
.##..##...
..#...#...
#####...#.
#..#.#.#.#
...#.#.#..
##.#...##.
..##.##.##
###.##.#..

Tile 2473:
#....####.
#..#.##...
#.##..#...
######.#.#
.#...#.#.#
.#########
.###.#..#.
########.#
##...##.#.
..###.#.#.

Tile 2971:
..#.#....#
#...###...
#.#.###...
##.##..#..
.#####..##
.#..####.#
#..#.#..#.
..####.###
..#.#.###.
...#.#.#.#

Tile 2729:
...#.#.#.#
####.#....
..#.#.....
....#..#.#
.##..##.#.
.#.####...
####.#.#..
##.####...
##..#.##..
#.##...##.

Tile 3079:
#.#.#####.
.#..######
..#.......
######....
####.#..#.
.#...#.##.
#.#####.##
..#.###...
..#.......
..#.###...";

        let test_arrangement = Arrangement::from_str(test_tiles_text);
        assert_eq!(count_water_roughness(&test_arrangement), 273);
    }
}
